### PR TITLE
Zos 600: Enable users to edit their profile

### PR DESCRIPTION
**Pull request by ratik21**
### What does this do?

Enables users to edit their profiles. 

SC:

image

image

image

Flow:

https://github.com/zer0-os/zOS/assets/33264364/0ae90a9c-770d-454c-956a-1ac703c7089b

### CODE REVIEW

File: src/components/settings-menu/index.test.tsx [hunk 1]
```diff
@@ -22,13 +23,23 @@ describe('settings-menu', () => {
     expect(wrapper).toHaveElement(DropdownMenu);
   });
 
+  it('calls onLogout prop when logout item is selected', function () {
```
Comment (dalefukami) [inline src/components/settings-menu/index.test.tsx:26]:
```suggestion
  it('opens profile dialog', function () {
```

File: src/store/edit-profile/index.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import { createAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
+
+export enum SagaActionTypes {
+  EditProfile = 'profile/edit',
+}
+
+export type EditProfileState = {
+  loading: boolean;
+  errors: string[];
+  changesSaved: boolean;
```
Comment (dalefukami) [inline src/store/edit-profile/index.ts:10]: :thinking: would the `loading` and `changesSaved` combo be better represented as a `state` enum? Like: `NONE`, `IN_PROGRES`, `SUCCESS` or something along those lines.

File: src/store/edit-profile/api.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  profileImage?: string;
+}) {
+  try {
+    const response = await patch(`/api/profiles/${profileId}`).send({ profileId, firstName: name, profileImage });
+    return {
+      success: true,
+      response: response.body,
+    };
+  } catch (error: any) {
+    if (error?.response?.status === 400) {
```
Comment (dalefukami) [inline src/store/edit-profile/api.ts:19]: Doesn't seem like we do anything with these errors in the saga so maybe we should just not catch the error here and let it throw. Did we confirm that errors for this endpoint come back in this format? (moot point if we remove this)

File: src/components/settings-menu/index.tsx [hunk 4]
```diff
@@ -66,39 +85,48 @@ export class SettingsMenu extends React.Component<Properties, State> {
...
+            {
+              id: 'header',
+              label: this.renderSettingsHeader(),
+              onSelect: () => {},
+            },
+            {
+              className: 'edit_profile',
+              id: 'edit_profile',
+              label: this.renderSettingsOption(<IconUser1 />, 'Edit Profile'),
+              onSelect: () => this.openEditProfileDialog(),
```
Comment (dalefukami) [inline src/components/settings-menu/index.tsx:101]:
```suggestion
              onSelect: this.openEditProfileDialog,
```

No need for the lambda. Might as well just pass the reference to the function.

**Review by dalefukami — APPROVED** [review]
Nice, looks pretty good. Just a few comments. I'll let you decide which ones should be handle before merging and which can be follow ups.

File: src/components/settings-menu/index.tsx [hunk 5]
```diff
@@ -66,39 +85,48 @@ export class SettingsMenu extends React.Component<Properties, State> {
...
-        trigger={
-          <Avatar
-            isActive={this.state.isDropdownOpen}
-            size={'medium'}
-            type={'circle'}
-            imageURL={this.props.userAvatarUrl}
-          />
-        }
-      />
+      <div>
```
[the remark this answers is not in the document]
Comment (ratik21) [reply src/components/settings-menu/index.tsx:88]: we don't. updated to fragment

File: src/store/authentication/saga.ts [hunk 6]
```diff
@@ -125,7 +126,22 @@ export function* clearUserState() {
   ]);
 }
 
+function* listenForUserProfileUpdates() {
+  const authChannel = yield call(getAuthChannel);
+  while (true) {
+    const { payload } = yield take(authChannel, Events.UserProfileUpdated);
```
[the remark this answers is not in the document]
Comment (ratik21) [reply src/store/authentication/saga.ts:132]: works! Sorry i forgot.

File: src/store/edit-profile/saga.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    if (response.success) {
+      yield call(publishUserProfileEdited, { name, profileImage });
+      return;
+    } else {
+      yield put(setErrors([ProfileDetailsErrors.UNKNOWN_ERROR]));
+    }
+  } catch (e) {
+    yield put(setErrors([ProfileDetailsErrors.UNKNOWN_ERROR]));
+  } finally {
+    yield put(setChangesSaved(true));
```
[the remark this answers is not in the document]
Comment (ratik21) [reply src/store/edit-profile/saga.ts:40]: makes sense. updated.

File: src/store/edit-profile/api.ts (hunk 3 above)
Comment (ratik21) [reply src/store/edit-profile/api.ts:19]: > Did we confirm that errors for this endpoint come back in this format?

We don't have specific 400 Errors for the PATCH endpoint, no. We can simply throw here.

File: src/components/edit-profile/container.tsx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+      currentProfileImage: user?.data?.profileSummary.profileImage,
+    };
+  }
+
+  static mapActions(_props: Properties): Partial<Properties> {
+    return { editProfile, setChangesSaved };
+  }
+
+  componentWillUnmount(): void {
+    this.props.setChangesSaved(false);
```
[the remark this answers is not in the document]
Comment (ratik21) [reply src/components/edit-profile/container.tsx:45]: done, thanks for suggestion.

File: src/store/edit-profile/index.ts (hunk 2 above)
Comment (ratik21) [reply src/store/edit-profile/index.ts:10]: done

File: src/store/authentication/channels.ts [hunk 9]
```diff
@@ -4,6 +4,7 @@ import { call } from 'redux-saga/effects';
 export enum Events {
   UserLogin = 'user/login',
   UserLogout = 'user/logout',
+  UserProfileUpdated = 'user/profile-updated',
```
[the remark this answers is not in the document]
Comment (ratik21) [reply src/store/authentication/channels.ts:7]: hmm, makes sense. I don't think any existing multicast channel is good enough. (Currently we have buses for `channels`/`conversations`, `eventChannels` for sendbird events, and the `authChannel` for authentication). We'll likely need to add a new bus i think. I'll discuss more async, & implement this one as a follow up.

**Pull request merged**